source/vsignal/sink.d

```diff
@@ -0,0 +1,199 @@
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, F))();
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+///
+Connection connect(alias pred, T, F)(auto ref Sink!F sink, ref T instance)
+{
+	sink.disconnect!pred(instance);
+
+	import core.lifetime : move;
+
+	with (sink)
+	{
+		Slot!F call;
+		call.slot_connect!pred(instance);
+		signal.calls ~= call.move();
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, T, F))(instance);
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+/++
+Disconnects a listener from a Signal. The order is not preserved after the
+removal.
+
+Examples:
+---
+void method() {}
+struct Foo { void method() {} }
+Foo foo;
+
+/* ... */
+
+// lets assume a signal exists with listeners
+sink.disconnect!(Foo.method)(foo); // disconnects Foo.method with payload foo
+sink.disconnect!method;            // disconnects method
+sink.disconnect(foo);              // disconnects all listeners with payload foo
+sink.disconnect();                 // disconnects all listeners
+---
+
+Params:
+	pred = the listener to disconnect.
+	sink = the Sink that holds the signal.
+	instance = the payload that composes the listener.
++/
+void disconnect(alias pred, F)(auto ref Sink!F sink)
+{
+	import std.algorithm : remove, SwapStrategy;
+	import std.algorithm : countUntil;
+
+	Slot!F call;
+	call.slot_connect!pred;
+
+	with (sink)
+	{
+		const index = signal.calls.countUntil(call);
+
+		if (index > -1)
+			signal.calls = signal.calls.remove!(SwapStrategy.unstable)(index);
+	}
+}
+
+///
+void disconnect(alias pred, T, F)(auto ref Sink!F sink, ref T instance)
+{
+	import std.algorithm : remove, SwapStrategy;
+	import std.algorithm : countUntil;
+
+	Slot!F call;
+	call.slot_connect!pred(instance);
+
+	with (sink)
+	{
+		const index = signal.calls.countUntil(call);
+
+		if (index > -1)
+			signal.calls = signal.calls.remove!(SwapStrategy.unstable)(index);
+	}
+}
+
+///
+void disconnect(T, F)(auto ref Sink!F sink, ref T instance)
+{
+	import std.algorithm : remove, SwapStrategy;
+
+	with (sink)
+	{
+		signal.calls = signal.calls.remove!(call => call.payload is &instance , SwapStrategy.unstable);
+	}
+}
+
+///
+void disconnect(F)(auto ref Sink!F sink)
+{
+	sinksignal.calls = [];
+}
+
+/**
+Used as a bridge between Sink and Connection to correctly disconnect a listener
+from a signal.
+
+Params:
+	pred = the listener to disconnect,
+	signal = the signal that contains the listener.
+	instance = the payload that is attached to the listener.
+*/
+private void release(alias pred, F)(void* signal)
+{
+	Sink!F(() @trusted { return cast(Signal!F*) signal; } ()).disconnect!pred();
+}
+
+///
+private void release(alias pred, T, F)(ref T instance, void* signal)
+{
+	Sink!F(() @trusted { return cast(Signal!F*) signal; } ()).disconnect!pred(instance);
+}
+
+struct Sink(F)
+{
+	this(ref Signal!F sig)
+	{
+		() @trusted { signal = &sig; } ();
```

Comment:
This is not fine to do. Currently, the compiler can't track references and this may escape.

Reply:
The variable `signal` is private. The user cannot access it therefore it cannot be escaped.

Reply:
`length` and `empty` deference `signal` and this can dereference an invalid stack frame leading to undefined behaviour